src/ui/handler/drag_pan.js

```diff
@@ -164,43 +180,82 @@ class DragPanHandler {
 
     _onMouseUp(e: MouseEvent) {
         if (DOM.mouseButton(e) !== 0) return;
-        this._finish(e);
+        switch (this._state) {
+        case 'active':
+            this._state = 'enabled';
+            DOM.suppressClick();
+            this._unbind();
+            this._deactivate();
+            this._inertialPan(e);
+            break;
+        case 'pending':
+            this._state = 'enabled';
+            this._unbind();
+            break;
+        default:
+            assert(false);
+            break;
+        }
     }
 
     _onTouchEnd(e: TouchEvent) {
-        this._finish(e);
+        switch (this._state) {
+        case 'active':
+            this._state = 'enabled';
+            this._unbind();
+            this._deactivate();
+            this._inertialPan(e);
+            break;
+        case 'pending':
+            this._state = 'enabled';
+            this._unbind();
+            break;
+        default:
+            assert(false);
+            break;
+        }
     }
 
     _onBlur(e: FocusEvent) {
-        this._finish(e);
+        switch (this._state) {
+        case 'active':
+            this._state = 'enabled';
+            this._unbind();
+            this._deactivate();
+            this._fireEvent('dragend', e);
+            this._fireEvent('moveend', e);
+            break;
+        case 'pending':
+            this._state = 'enabled';
+            this._unbind();
+            break;
+        default:
+            assert(false);
+            break;
+        }
     }
 
-    _finish(e: MouseEvent | TouchEvent | FocusEvent) {
+    _unbind() {
         window.document.removeEventListener('touchmove', this._onMove, {capture: true});
         window.document.removeEventListener('touchend', this._onTouchEnd);
         window.document.removeEventListener('mousemove', this._onMove, {capture: true});
         window.document.removeEventListener('mouseup', this._onMouseUp);
         window.removeEventListener('blur', this._onBlur);
+    }
 
-        if (!this.isActive()) return;
-
-        this._active = false;
+    _deactivate() {
         delete this._lastMoveEvent;
         delete this._previousPos;
         delete this._pos;
+    }
```

Comment:
Just noting: because there's a state called `'active'`, I (briefly) found it confusing to have a method called `_deactivate` that doesn't affect `this.state`... but I can't think of any better names.